lib/compiler/src/engine/artifact.rs

```diff
@@ -378,47 +392,81 @@ impl ArtifactCreate for Artifact {
+    }
+
+    fn internal_register_frame_info(&mut self) {
+        if self
             .allocated
             .as_ref()
             .expect("It must be allocated")
-            .frame_info_registration
-            .as_ref()
+            .frame_info_registered
         {
-            let mut info = frame_info_registration.lock().unwrap();
+            return; // already done
+        }
 
-            if info.is_some() {
-                return;
-            }
+        let finished_function_extents = self
+            .allocated
+            .as_ref()
+            .expect("It must be allocated")
+            .finished_functions
+            .values()
+            .copied()
+            .zip(
+                self.allocated
+                    .as_ref()
+                    .expect("It must be allocated")
+                    .finished_function_lengths
+                    .values()
+                    .copied(),
+            )
+            .map(|(ptr, length)| FunctionExtent { ptr, length })
+            .collect::<PrimaryMap<LocalFunctionIndex, _>>()
+            .into_boxed_slice();
 
-            let finished_function_extents = self
-                .allocated
-                .as_ref()
-                .expect("It must be allocated")
-                .finished_functions
-                .values()
-                .copied()
-                .zip(
-                    self.allocated
-                        .as_ref()
-                        .expect("It must be allocated")
-                        .finished_function_lengths
-                        .values()
-                        .copied(),
-                )
-                .map(|(ptr, length)| FunctionExtent { ptr, length })
-                .collect::<PrimaryMap<LocalFunctionIndex, _>>()
-                .into_boxed_slice();
-
-            let frame_infos = self.artifact.get_frame_info_ref();
-            *info = register_frame_info(
-                self.artifact.create_module_info(),
-                &finished_function_extents,
-                frame_infos.clone(),
-            );
-        }
+        let frame_info_registration = &mut self
+            .allocated
+            .as_mut()
+            .expect("It must be allocated")
+            .frame_info_registration;
+
+        let frame_infos = self.artifact.get_frame_info_ref();
+
+        *frame_info_registration = register_frame_info(
+            self.artifact.create_module_info(),
+            &finished_function_extents,
+            frame_infos.clone(),
+        );
+
+        self.allocated
+            .as_mut()
+            .expect("It must be allocated")
+            .frame_info_registered = true;
+    }
+
+    /// The GlobalFrameInfoRegistration needs to be transfered to EngineInner if
+    /// register_frame_info has been used.
+    #[deprecated(since = "4.0.0", note = "done automaticaly by Artifact::from_parts.")]
+    pub fn transfert_frame_info_registration(&mut self) -> Option<GlobalFrameInfoRegistration> {
```

Comment:
Do you mean `transfer_frame_info_registration` ? (there's an extra `t`)

Reply:
I think we can name it `take_frame_info_registration`

Reply:
oh, spelling...

Reply:
fixed